### PR TITLE
OS-274: Added Search functionality in Discussion forum

### PR DESCRIPTION
**Pull request by HovGrigoryan**

### CODE REVIEW

File: os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java [hunk 1]
```diff
@@ -301,11 +304,12 @@ public ResponseEntity<MentorQuestionResponseDto> createMentorQuestion(
   @GetMapping("/{enrolledCourseId}/mentor-questions")
   public ResponseEntity<Page<MentorQuestionResponseDto>> findMentorQuestionsByCourseId(
       @Parameter(description = "Enrolled course id") @PathVariable Long enrolledCourseId,
-      Pageable pageable) {
+      Pageable pageable,
+      @Valid @Size(min = 3) @RequestParam(value = "q", required = false) String searchQuery) {
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java:308]: please define a custom validation message for this case

File: os-server/src/test/java/app/openschool/course/discussion/question/MentorQuestionServiceImplTest.java [hunk 2]
```diff
@@ -78,33 +78,37 @@ void findMentorQuestionByCourseId_withCorrectData() {
...
     verify(mentorQuestionRepository, times(1))
-        .findQuestionByEnrolledCourseId(enrolledCourseId, pageable);
+        .findQuestionByEnrolledCourseId(enrolledCourseId, pageable, searchQuery);
   }
 
   @Test
   void findMentorQuestionByCourseId_withIncorrectData() {
     long wrongEnrolledCourseId = 1L;
     Pageable pageable = PageRequest.of(0, 2);
+    String searchQuery = "";
```
Comment (SevakMart) [inline os-server/src/test/java/app/openschool/course/discussion/question/MentorQuestionServiceImplTest.java:101]: you didn't covered case with null

**Review by SevakMart — CHANGES_REQUESTED** [review]
please fix/answer to the comments

File: os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java [hunk 3]
```diff
@@ -157,11 +159,12 @@ public ResponseEntity<HttpStatus> deletePeersQuestion(
   @GetMapping("/{enrolledCourseId}/peers-questions")
   public ResponseEntity<Page<QuestionResponseDto>> findQuestionsByCourseId(
       @Parameter(description = "Enrolled course id") @PathVariable Long enrolledCourseId,
-      Pageable pageable) {
+      Pageable pageable,
+      @Valid @Size(min = 3) @RequestParam(value = "q", required = false) String searchQuery) {
```
[the remark this answers is not in the document]
Comment (HovGrigoryan) [reply os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java:163]: Yes, the validation allows for a null value. It has been tested and confirmed to work as expected.

File: os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java (hunk 1 above)
Comment (HovGrigoryan) [reply os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java:311]: Done✅

File: os-server/src/test/java/app/openschool/course/discussion/question/MentorQuestionServiceImplTest.java (hunk 2 above)
Comment (HovGrigoryan) [reply os-server/src/test/java/app/openschool/course/discussion/question/MentorQuestionServiceImplTest.java:101]: Done✅

File: os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java [hunk 4]
```diff
@@ -157,11 +159,15 @@ public ResponseEntity<HttpStatus> deletePeersQuestion(
   @GetMapping("/{enrolledCourseId}/peers-questions")
   public ResponseEntity<Page<QuestionResponseDto>> findQuestionsByCourseId(
       @Parameter(description = "Enrolled course id") @PathVariable Long enrolledCourseId,
-      Pageable pageable) {
+      Pageable pageable,
+      @Valid
+          @Size(min = 3, message = "Search query must be at least 3 characters long")
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java:164]: this should be in the properties file, to be able to easily translate

File: os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java [hunk 5]
```diff
@@ -24,17 +24,20 @@
...
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.PutMapping;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 @RequestMapping("/api/v1/courses/enrolled")
+@Validated
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/course/discussion/DiscussionForumController.java:40]: please remove this one, as it is not needed anymore

File: os-server/src/main/java/app/openschool/course/discussion/peers/question/PeersQuestionServiceImpl.java [hunk 6]
```diff
@@ -68,8 +75,15 @@ public void delete(Long questionId, Long enrolledCourseId, String currentUserEma
   }
 
   @Override
-  public Page<PeersQuestion> findQuestionByCourseId(Long enrolledCourseId, Pageable pageable) {
-    return peersQuestionRepository.findQuestionByEnrolledCourseId(enrolledCourseId, pageable);
+  public Page<PeersQuestion> findQuestionByCourseId(
+      Long enrolledCourseId, Pageable pageable, String searchQuery) {
+    if (searchQuery != null && searchQuery.length() < 3) {
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/course/discussion/peers/question/PeersQuestionServiceImpl.java:80]: please use Objects.nonNull method, and move the validation logic into a separate private method

Comment (SevakMart) [reply os-server/src/main/java/app/openschool/course/discussion/peers/question/PeersQuestionServiceImpl.java:80]: make the number a constant

**Review by SevakMart — APPROVED** [review]

**Pull request merged**